File: src/test/java/me/nasukhov/bot/TukitaLearner/command/AskQuestionTest.java
package me.nasukhov.bot.TukitaLearner.command;

import me.nasukhov.TukitaLearner.bot.command.AskQuestion;
import me.nasukhov.TukitaLearner.bot.io.*;
import me.nasukhov.TukitaLearner.study.GroupQuestion;
import me.nasukhov.TukitaLearner.study.ProgressRepository;
import me.nasukhov.TukitaLearner.study.Question;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Optional;
import java.util.UUID;

import static org.mockito.Mockito.*;

public class AskQuestionTest {
    private static final String TEMPLATE_ASK_QUESTION = "qh answer %s %d";
    private AskQuestion handler;
    private ProgressRepository progressRepository;
    private Output output;

    @BeforeEach
    void setup() {
        output = mock(Output.class);
        progressRepository = mock(ProgressRepository.class);

        handler = new AskQuestion(progressRepository);
    }

    @ParameterizedTest
    @ValueSource(strings = {"/ask", "/ask@botName"})
    void testHandleAsk(String rawInput) {
        UUID chQuestionId = UUID.randomUUID();

        GroupQuestion groupQuestion = new GroupQuestion(
                chQuestionId,
                new Question(123, "2+5 equals to", "seven", new ArrayList<>(){{
                    add("thirteen");
                    add("seven");
                    add("twenty-one");
                }})
        );

        when(progressRepository.createRandomForChannel("SomeChannelId")).thenReturn(Optional.of(groupQuestion));

        Input input = new Input(
                rawInput,
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );
        handler.handle(input, output);

        verify(output, only()).promptChoice("2+5 equals to", new HashMap<>() {{
            put("thirteen", String.format(TEMPLATE_ASK_QUESTION, chQuestionId, 1));
            put("seven", String.format(TEMPLATE_ASK_QUESTION, chQuestionId, 2));
            put("twenty-one", String.format(TEMPLATE_ASK_QUESTION, chQuestionId, 3));
        }});
    }
}

File: src/test/kotlin/me/nasukhov/tukitalearner/bot/AskQuestionTest.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.io.Channel
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.PromptOutput
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.bot.io.User
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.GroupQuestionRepository
import me.nasukhov.tukitalearner.study.GroupRepository
import me.nasukhov.tukitalearner.study.Question
import me.nasukhov.tukitalearner.study.QuestionRepository
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertInstanceOf
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@SpringBootTest
@Transactional
class AskQuestionTest(
    @Autowired private val handler: AskQuestion,
    @Autowired private val groupQuestionRepository: GroupQuestionRepository,
    @Autowired private val questionRepository: QuestionRepository,
    @Autowired private val groupRepository: GroupRepository,
) {
    private lateinit var group: Group

    companion object {
        private const val TEMPLATE_ASK_QUESTION = "qh answer %s %d"
    }

    @BeforeEach
    fun setup() {
        group = Group("SomeGroup123")
        groupRepository.save(group)
        questionRepository.save(
            Question(
                "2+5 equals to",
                "seven",
                listOf(
                    "thirteen",
                    "seven",
                    "twenty-one",
                ),
            ),
        )
    }

    @AfterEach
    fun reset() {
        questionRepository.deleteAll()
    }

    @Test
    fun handleAskWhenAllQuestionsAreAnswered() {
        val channel = Channel(group.id)
        val user = User("SomeId", "SomeName")
        val input = Input("/ask", channel, user)

        questionRepository.deleteAll()

        val output = handler.handle(input)
        assertInstanceOf<TextOutput>(output)
        assertEquals("У нас пока нет новых вопросов. Проверьте позже", output.value)

        assertTrue(groupQuestionRepository.findAll().isEmpty())
    }

    @ParameterizedTest
    @ValueSource(strings = ["/ask", "/ask@botName"])
    fun handleAsk(rawInput: String) {
        val channel = Channel(group.id)
        val user = User("SomeId", "SomeName")
        val input = Input(rawInput, channel, user)

        assertTrue(groupQuestionRepository.findAll().isEmpty())

        val output = handler.handle(input)
        val allQuestions = groupQuestionRepository.findAll()
        val groupQuestion = allQuestions.first()

        assertInstanceOf<PromptOutput>(output)
        assertEquals(
            PromptOutput(
                "2+5 equals to",
                mapOf(
                    "thirteen" to String.format(TEMPLATE_ASK_QUESTION, groupQuestion.id, 1),
                    "seven" to String.format(TEMPLATE_ASK_QUESTION, groupQuestion.id, 2),
                    "twenty-one" to String.format(TEMPLATE_ASK_QUESTION, groupQuestion.id, 3),
                ),
            ),
            output,
        )

        assertEquals(1, allQuestions.size, "Expected one question")
        assertEquals("2+5 equals to", groupQuestion.text)
        Assertions.assertArrayEquals(
            arrayOf("thirteen", "seven", "twenty-one"),
            groupQuestion.listVariants().toTypedArray(),
        )
    }
}
